src/reply_parser.c

```diff
@@ -0,0 +1,236 @@
+    const char *proto = parser->curr_location;
+    char *p = strchr(proto+1,'\r');
+
+    parser->curr_location = p + 2; /* for \r\n */
+
+    parser->simple_str_callback(p_ctx, proto+1, p-proto-1, proto, parser->curr_location - proto);
+
+    return C_OK;
+}
+
+static int parseError(ReplyParser* parser, void* p_ctx) {
+    const char *proto = parser->curr_location;
+    char *p = strchr(proto+1,'\r');
+
+    parser->curr_location = p + 2; // for \r\n
+
+    parser->error_callback(p_ctx, proto+1, p-proto-1, proto, parser->curr_location - proto);
+
+    return C_OK;
+}
+
+static int parseLong(ReplyParser* parser, void* p_ctx) {
+    const char *proto = parser->curr_location;
+    char *p = strchr(proto+1,'\r');
+
+    parser->curr_location = p + 2; /* for \r\n */
+
+    long long val;
+    string2ll(proto+1,p-proto-1,&val);
+
+    parser->long_callback(p_ctx, val, proto, parser->curr_location - proto);
+
+    return C_OK;
+}
+
+static int parseAttributes(ReplyParser* parser, void* p_ctx) {
+    const char *proto = parser->curr_location;
+    char *p = strchr(proto+1,'\r');
+    long long len;
+
+    string2ll(proto+1,p-proto-1,&len);
+    p += 2;
+
+    parser->curr_location = p;
+
+    parser->attribute_callback(parser, p_ctx, len, proto);
+
+    return C_OK;
+}
+
+static int parseVerbatimString(ReplyParser* parser, void* p_ctx) {
+    const char *proto = parser->curr_location;
+    char *p = strchr(proto+1,'\r');
+    long long bulklen;
+
+
+    parser->curr_location = p + 2; /* for \r\n */
+
+    string2ll(proto+1,p-proto-1,&bulklen);
+    const char* format = parser->curr_location;
+    parser->curr_location += bulklen;
+
+    parser->curr_location += 2; /* for \r\n */
+
+    parser->verbatim_string_callback(p_ctx, format, format + 4, bulklen - 4, proto, parser->curr_location - proto);
+
+    return C_OK;
+}
+
+static int parseBigNumber(ReplyParser* parser, void* p_ctx) {
+    const char *proto = parser->curr_location;
+    char *p = strchr(proto+1,'\r');
+
+    parser->curr_location = p + 2; /* for \r\n */
+
+    parser->big_number_callback(p_ctx, proto+1, p-proto-1, proto, parser->curr_location - proto);
+
+    return C_OK;
+}
+
+static int parseNull(ReplyParser* parser, void* p_ctx) {
+    const char *proto = parser->curr_location;
+    char *p = strchr(proto+1,'\r');
+
+    parser->curr_location = p + 2; /* for \r\n */
+
+    parser->null_callback(p_ctx, proto, parser->curr_location - proto);
+
+    return C_OK;
+}
+
+static int parseDouble(ReplyParser* parser, void* p_ctx) {
+    const char *proto = parser->curr_location;
+    char *p = strchr(proto+1,'\r');
+    parser->curr_location = p + 2; /* for \r\n */
+    char buf[MAX_LONG_DOUBLE_CHARS+1];
+    size_t len = p-proto-1;
+    double d;
+
+    if (len <= MAX_LONG_DOUBLE_CHARS) {
+        memcpy(buf,proto+1,len);
+        buf[len] = '\0';
+        d = strtod(buf,NULL); /* We expect a valid representation. */
+    } else {
+        d = 0;
```

Comment:
maybe we need a way to propagate some error in case strtod fails, or the string is too long?

Reply:
The idea in this parser is that it does not need to handle replies error as it parse replies generated by Redis, this allowes it to be faster. You think we should handle errors anyway? (We will probably lose performance)

Reply:
I'm not sure. If the previous ones didn't handle it, and it's an api that's internal to redis (we are free to change it when we like), then we can skip that now

Reply:
The previous ones didn't handle it, and its also clearly stated on the Lua parser today: https://github.com/redis/redis/blob/ac8b1df8850cc80fbf9ce8c2fbde0c1d3a1b4e91/src/scripting.c#L124

Reply:
This kind of assumptions back fire over time, I think we need a really good reason (backed by data) to create a fragile parser.